File: kotlin/rayfold-core/src/test/kotlin/dev/rayfold/core/Fixture.kt
package dev.rayfold.core

import kotlinx.coroutines.flow.asFlow
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** Declarative resolver interpreter for conformance fixtures (mirrors conformance/src/fixture.ts). */
class FixtureStore(data: JsonObject) {
    val tables: MutableMap<String, MutableList<MutableMap<String, JsonElement>>> =
        data.mapValues { (_, rows) -> rows.jsonArray.map { it.jsonObject.toMutableMap() }.toMutableList() }.toMutableMap()
    val calls = linkedMapOf<String, Int>()
    var nextId = 1
    fun table(name: String) = tables.getOrPut(name) { mutableListOf() }
    fun count(k: String) { calls[k] = (calls[k] ?: 0) + 1 }
}

object FixtureResolvers {
    private class Env(val args: JsonElement, val viewer: JsonElement, val row: JsonElement? = null)

    private fun JsonObject.req(key: String): JsonElement = this[key] ?: error("fixture resolver spec is missing \"$key\"")
    private fun JsonObject.str(key: String): String = req(key).jsonPrimitive.content

    private fun tmpl(v: JsonElement?, env: Env): JsonElement = when (v) {
        null -> JsonNull
        is JsonPrimitive -> {
            val s = v.takeIf { it.isString }?.content
            if (s != null && s.startsWith("$")) {
                val parts = s.substring(1).split(".")
                var cur: JsonElement? = when (parts[0]) { "args" -> env.args; "viewer" -> env.viewer; "row" -> env.row; else -> null }
                for (p in parts.drop(1)) cur = (cur as? JsonObject)?.get(p)
                cur ?: JsonNull
            } else v
        }
        is JsonArray -> JsonArray(v.map { tmpl(it, env) })
        is JsonObject -> JsonObject(v.mapValues { (_, x) -> tmpl(x, env) })
    }

    private fun matches(row: Map<String, JsonElement>, where: JsonObject?, env: Env): Boolean {
        if (where == null) return true
        return where.all { (k, v) ->
            val want = tmpl(v, env)
            want is JsonNull || (row[k] as? JsonPrimitive)?.content == (want as? JsonPrimitive)?.content
        }
    }

    private fun page(rows: List<Map<String, JsonElement>>, p: JsonObject?): JsonObject {
        val first = (p?.get("first") as? JsonPrimitive)?.contentOrNull?.toDoubleOrNull()?.toInt() ?: 20
        val after = (p?.get("after") as? JsonPrimitive)?.takeIf { it.isString }?.content
        var start = 0
        if (after != null) { val i = rows.indexOfFirst { (it["id"] as? JsonPrimitive)?.content == after }; start = if (i < 0) rows.size else i + 1 }
        val items = rows.drop(start).take(first)
        return buildJsonObject {
            put("items", JsonArray(items.map { JsonObject(it) }))
            put("cursor", items.lastOrNull()?.get("id") ?: JsonNull)
            put("hasMore", JsonPrimitive(start + first < rows.size))
            put("total", JsonPrimitive(rows.size))
        }
    }

    fun build(fixture: JsonObject, store: FixtureStore): Resolvers {
        val res = fixture.req("resolvers").jsonObject
        val queries = (res["Query"]?.jsonObject ?: JsonObject(emptyMap())).mapValues { (name, specE) ->
            val spec = specE.jsonObject
            val fn: RootResolver = { args, ctx ->
                store.count("Query.$name")
                val env = Env(args, ctx.viewer)
                val rows = store.table(spec.str("from")).filter { matches(it, spec["where"] as? JsonObject, env) }
                when (spec.str("mode")) {
                    "one" -> rows.firstOrNull()?.let { JsonObject(it) } ?: JsonNull
                    "list" -> JsonArray(rows.map { JsonObject(it) })
                    else -> page(rows, args[(spec["pageArg"] as? JsonPrimitive)?.content ?: "page"] as? JsonObject)
                }
            }
            fn
        }
        val fields = (res["fields"]?.jsonObject ?: JsonObject(emptyMap())).mapValues { (type, fieldsE) ->
            fieldsE.jsonObject.mapValues { (field, specE) ->
                val spec = specE.jsonObject
                val loader: FieldLoader = { parents, args, _ ->
                    store.count("$type.$field")
                    val rows = store.table(spec.str("from"))
                    val key = spec.str("key")
                    val match = spec.str("match")
                    parents.map { p ->
                        val hits = rows.filter { (it[match] as? JsonPrimitive)?.content == (p[key] as? JsonPrimitive)?.content }
                        when (spec.str("mode")) {
                            "one" -> hits.firstOrNull()?.let { JsonObject(it) }
                            "list" -> JsonArray(hits.map { JsonObject(it) })
                            else -> page(hits, args[(spec["pageArg"] as? JsonPrimitive)?.content ?: "page"] as? JsonObject)
                        }
                    }
                }
                loader
            }
        }
        val commands = (res["Command"]?.jsonObject ?: JsonObject(emptyMap())).mapValues { (name, specE) ->
            val spec = specE.jsonObject
            val fn: suspend (JsonObject, RayfoldContext) -> Any? = { args, ctx ->
                store.count("Command.$name")
                val env = Env(args, ctx.viewer)
                for (failE in spec["fail"]?.jsonArray ?: JsonArray(emptyList())) {
                    val fail = failE.jsonObject
                    val w = fail.req("when").jsonObject
                    val from = (w["from"] as? JsonPrimitive)?.content
                    val subject = if (from != null) store.table(from).firstOrNull { matches(it, w["where"] as? JsonObject, env) } else null
                    val field = w.str("field")
                    val actual = if (subject != null) subject[field] else tmpl(JsonPrimitive("\$args.$field"), env)
                    val rowEnv = Env(args, ctx.viewer, subject?.let { JsonObject(it) })
                    val want = tmpl(w["value"], rowEnv)
                    val hit = when (w.str("op")) {
                        "missing" -> subject == null
                        "eq" -> (actual as? JsonPrimitive)?.content == (want as? JsonPrimitive)?.content
                        else -> ((actual as? JsonPrimitive)?.content?.toDoubleOrNull() ?: 0.0) < ((want as? JsonPrimitive)?.content?.toDoubleOrNull() ?: 0.0)
                    }
                    if (hit) {
                        val data = tmpl(fail["data"] ?: JsonObject(emptyMap()), rowEnv)
                        val type = (fail["type"] as? JsonPrimitive)?.content
                        val message = (fail["message"] as? JsonPrimitive)?.content
                        if (type != null) throw RayfoldException.domain(type, data, message ?: type)
                        val code = (fail["code"] as? JsonPrimitive)?.content ?: "failed_precondition"
                        throw RayfoldException(Code.entries.first { it.wire == code }, message ?: code)
                    }
                }
                var inserted: MutableMap<String, JsonElement>? = null
                var updated: MutableMap<String, JsonElement>? = null
                (spec["update"] as? JsonObject)?.let { u ->
                    val row = store.table(u.str("table")).firstOrNull { matches(it, u["where"] as? JsonObject, env) }
                        ?: throw RayfoldException(Code.NOT_FOUND, "row not found")
                    val versionField = (u["version"] as? JsonPrimitive)?.content
                    if (versionField != null) ctx.checkVersion("${u.str("table")}:${(row["id"] as JsonPrimitive).content}", row[versionField], JsonObject(row))
                    val target = if (ctx.simulate) row.toMutableMap() else row
                    (u["merge"])?.let { m -> (tmpl(m, env) as? JsonObject)?.forEach { (k, v) -> target[k] = v } }
                    if (versionField != null) target[versionField] = num(((target[versionField] as? JsonPrimitive)?.content?.toDoubleOrNull() ?: 0.0) + 1)
                    for ((k, v) in (u["set"] as? JsonObject ?: JsonObject(emptyMap()))) {
                        val add = (v as? JsonObject)?.get("\$add"); val sub = (v as? JsonObject)?.get("\$sub")
                        val cur = (target[k] as? JsonPrimitive)?.content?.toDoubleOrNull() ?: 0.0
                        target[k] = when {
                            add != null -> num(cur + (tmpl(add, env) as JsonPrimitive).content.toDouble())
                            sub != null -> num(cur - (tmpl(sub, env) as JsonPrimitive).content.toDouble())
                            else -> tmpl(v, env)
                        }
                    }
                    updated = target
                }
                (spec["insert"] as? JsonObject)?.let { ins ->
                    val prefix = (ins["idPrefix"] as? JsonPrimitive)?.content ?: "n"
                    val row = linkedMapOf<String, JsonElement>("id" to JsonPrimitive("$prefix${store.nextId++}"))
                    row.putAll(tmpl(ins["row"], env).jsonObject)
                    if (!ctx.simulate) store.table(ins.str("into")).add(row)
                    inserted = row
                }
                val result = if (spec.str("returns") == "inserted") inserted else updated
                val rowEnv = Env(args, ctx.viewer, result?.let { JsonObject(it) })
                val patch = (spec["patch"] as? JsonArray)?.map { tmpl(it, rowEnv).jsonObject } ?: emptyList()
                val emit = (spec["emit"] as? JsonArray)?.map { e -> e.jsonObject.str("event") to tmpl(e.jsonObject["payload"], rowEnv).jsonObject } ?: emptyList()
                CommandResult(result?.let { JsonObject(it) }, patch, emit)
            }
            fn
        }
        val streams = (res["Stream"]?.jsonObject ?: JsonObject(emptyMap())).mapValues { (name, specE) ->
            val items = specE.jsonObject.req("items").jsonArray.toList()
            val fn: StreamResolver = { _, _ -> store.count("Stream.$name"); items.asFlow() }
            fn
        }
        return Resolvers(queries, commands, streams, fields)
    }

    private fun num(d: Double): JsonPrimitive = if (d == Math.floor(d) && !d.isInfinite()) JsonPrimitive(d.toLong()) else JsonPrimitive(d)
}

File: conformance/src/fixture.ts
/**
 * Conformance fixture format and the declarative resolver interpreter.
 * Both runtimes (TS, Kotlin) implement this interpreter so fixtures need no code.
 */
import { RayfoldError, ok, type Frame, type PatchOp, type RayfoldContext, type RequestEnvelope, type Resolvers } from "@rayfold/server";

export interface Fixture {
  name: string;
  /** `.rayfold` schema text */
  schema: string;
  /** rows per type name; re-seeded before every case */
  data: Record<string, Array<Record<string, unknown>>>;
  resolvers: {
    Query?: Record<string, QuerySpec>;
    Command?: Record<string, CommandSpec>;
    Stream?: Record<string, StreamSpec>;
    fields?: Record<string, Record<string, FieldSpec>>; // "Type": { "field": spec }
  };
  options?: { budget?: number; maxDepth?: number; trustedShapes?: boolean; registerShapes?: string[] };
  cases: Case[];
}

export interface Case {
  name: string;
  viewer?: unknown;
  request: RequestEnvelope;
  /** Expected frames. Frames of different ops may interleave; per-op order must match. */
  frames: Frame[];
  /** Expected loader call counts, e.g. { "Book.author": 1 } */
  calls?: Record<string, number>;
  /** Run this many times in sequence against the same store (idempotency cases). */
  repeat?: number;
}

/** Value templates: "$args.x.y", "$viewer.id", "$parent.field", "$row.field", literals. */
type Tmpl = unknown;

export interface QuerySpec {
  from: string;
  where?: Record<string, Tmpl>;
  /** "one" returns the first match or null; "page" returns a Page from $args.page; "list" returns all matches */
  mode: "one" | "page" | "list";
  pageArg?: string; // default "page"
}

export interface FieldSpec {
  from: string;
  /** parent field to match against row field: parent[key] == row[match] */
  key: string;
  match: string;
  mode: "one" | "page" | "list";
  pageArg?: string;
}

export interface CommandSpec {
  /** conditions evaluated in order; the first one that matches throws */
  fail?: Array<{ when: { field: string; op: "lt" | "eq" | "missing"; value?: Tmpl; from?: string; where?: Record<string, Tmpl> }; type?: string; code?: string; data?: Record<string, Tmpl>; message?: string }>;
  insert?: { into: string; row: Record<string, Tmpl>; idPrefix?: string };
  /** `merge` copies the keys present in a template object (partial updates); `version` names a @version field to check and bump */
  update?: { table: string; where: Record<string, Tmpl>; set?: Record<string, Tmpl | { $add: Tmpl } | { $sub: Tmpl }>; merge?: Tmpl; version?: string };
  /** which row to return: "inserted" | "updated" */
  returns: "inserted" | "updated";
  patch?: PatchOp[];
  emit?: Array<{ event: string; payload: Record<string, Tmpl> }>;
}

export interface StreamSpec {
  /** yields these items in order, then ends */
  items: Array<Record<string, unknown>>;
}

export interface FixtureStore {
  tables: Record<string, Array<Record<string, unknown>>>;
  calls: Record<string, number>;
  nextId: number;
}

export function seedStore(f: Fixture): FixtureStore {
  return { tables: structuredClone(f.data), calls: {}, nextId: 1 };
}

function tmpl(v: Tmpl, env: { args?: unknown; viewer?: unknown; parent?: unknown; row?: unknown }): unknown {
  if (typeof v === "string" && v.startsWith("$")) {
    const [root, ...path] = v.slice(1).split(".");
    let cur: unknown = (env as Record<string, unknown>)[root!];
    for (const p of path) cur = cur && typeof cur === "object" ? (cur as Record<string, unknown>)[p] : undefined;
    return cur === undefined ? null : cur;
  }
  if (Array.isArray(v)) return v.map((x) => tmpl(x, env));
  if (v && typeof v === "object") {
    const o: Record<string, unknown> = {};
    for (const [k, x] of Object.entries(v)) o[k] = tmpl(x, env);
    return o;
  }
  return v;
}

function matches(row: Record<string, unknown>, where: Record<string, Tmpl> | undefined, env: Parameters<typeof tmpl>[1]): boolean {
  if (!where) return true;
  return Object.entries(where).every(([k, v]) => {
    const want = tmpl(v, env);
    return want === null ? true : String(row[k]) === String(want);
  });
}

function page(rows: Array<Record<string, unknown>>, p: { first: number; after: string | null } | undefined) {
  const first = p?.first ?? 20;
  let start = 0;
  if (p?.after) {
    const i = rows.findIndex((r) => String(r["id"]) === p.after);
    start = i < 0 ? rows.length : i + 1;
  }
  const items = rows.slice(start, start + first);
  return { items, cursor: items.length ? String(items[items.length - 1]!["id"]) : null, hasMore: start + first < rows.length, total: rows.length };
}

function count(store: FixtureStore, k: string): void {
  store.calls[k] = (store.calls[k] ?? 0) + 1;
}

export function fixtureResolvers(f: Fixture, store: FixtureStore): Resolvers {
  const r: Resolvers = { Query: {}, Command: {}, Stream: {} };
  const table = (name: string) => store.tables[name] ?? (store.tables[name] = []);

  for (const [name, spec] of Object.entries(f.resolvers.Query ?? {})) {
    r.Query![name] = (args: Record<string, unknown>, ctx: RayfoldContext<never>) => {
      count(store, `Query.${name}`);
      const env = { args, viewer: ctx.viewer };
      const rows = table(spec.from).filter((row) => matches(row, spec.where, env));
      if (spec.mode === "one") return rows[0] ?? null;
      if (spec.mode === "list") return rows;
      return page(rows, args[spec.pageArg ?? "page"] as never);
    };
  }
  for (const [type, fields] of Object.entries(f.resolvers.fields ?? {})) {
    r[type] = {};
    for (const [field, spec] of Object.entries(fields)) {
      r[type]![field] = ((parents: Array<Record<string, unknown>>, args: Record<string, unknown>) => {
        count(store, `${type}.${field}`);
        const rows = table(spec.from);
        return parents.map((p) => {
          const hits = rows.filter((row) => String(row[spec.match]) === String(p[spec.key]));
          if (spec.mode === "one") return hits[0] ?? null;
          if (spec.mode === "list") return hits;
          return page(hits, args[spec.pageArg ?? "page"] as never);
        });
      }) as never;
    }
  }
  for (const [name, spec] of Object.entries(f.resolvers.Command ?? {})) {
    r.Command![name] = (args: Record<string, unknown>, ctx: RayfoldContext<never>) => {
      count(store, `Command.${name}`);
      const env = { args, viewer: ctx.viewer };
      for (const fail of spec.fail ?? []) {
        const w = fail.when;
        let subject: Record<string, unknown> | undefined;
        if (w.from) subject = table(w.from).find((row) => matches(row, w.where, env));
        const actual = subject ? subject[w.field] : tmpl(`$args.${w.field}`, env);
        const want = tmpl(w.value, { ...env, row: subject });
        const hit = w.op === "missing" ? subject === undefined : w.op === "eq" ? String(actual) === String(want) : Number(actual) < Number(want);
        if (hit) {
          const data = tmpl(fail.data ?? {}, { ...env, row: subject }) as Record<string, unknown>;
          if (fail.type) throw RayfoldError.domain(fail.type, data, fail.message ?? fail.type);
          throw new RayfoldError((fail.code ?? "failed_precondition") as never, fail.message ?? fail.code ?? "failed");
        }
      }
      let inserted: Record<string, unknown> | undefined;
      let updated: Record<string, unknown> | undefined;
      if (spec.update) {
        updated = table(spec.update.table).find((row) => matches(row, spec.update!.where, env));
        if (!updated) throw new RayfoldError("not_found", "row not found");
        const vf = spec.update.version;
        if (vf) ctx.checkVersion(`${spec.update.table}:${String(updated["id"])}`, updated[vf], { ...updated });
        const target = ctx.simulate ? { ...updated } : updated;
        if (spec.update.merge !== undefined) {
          const src = tmpl(spec.update.merge, env);
          if (src && typeof src === "object" && !Array.isArray(src)) Object.assign(target, src);
        }
        if (vf) target[vf] = Number(target[vf]) + 1;
        for (const [k, v] of Object.entries(spec.update.set ?? {})) {
          if (v && typeof v === "object" && "$add" in (v as object)) target[k] = Number(target[k]) + Number(tmpl((v as { $add: Tmpl }).$add, env));
          else if (v && typeof v === "object" && "$sub" in (v as object)) target[k] = Number(target[k]) - Number(tmpl((v as { $sub: Tmpl }).$sub, env));
          else target[k] = tmpl(v, env);
        }
        updated = target;
      }
      if (spec.insert) {
        inserted = { id: `${spec.insert.idPrefix ?? "n"}${store.nextId++}`, ...(tmpl(spec.insert.row, env) as Record<string, unknown>) };
        if (!ctx.simulate) table(spec.insert.into).push(inserted);
      }
      const result = spec.returns === "inserted" ? inserted : updated;
      const rowEnv = { ...env, row: result };
      const extra: { patch?: PatchOp[]; emit?: Array<{ event: string; payload: Record<string, unknown> }> } = {};
      if (spec.patch) extra.patch = tmpl(spec.patch, rowEnv) as PatchOp[];
      if (spec.emit) extra.emit = spec.emit.map((e) => ({ event: e.event, payload: tmpl(e.payload, rowEnv) as Record<string, unknown> }));
      return ok(result, extra);
    };
  }
  for (const [name, spec] of Object.entries(f.resolvers.Stream ?? {})) {
    r.Stream![name] = () => {
      count(store, `Stream.${name}`);
      return (async function* () {
        for (const it of spec.items) yield it;
      })();
    };
  }
  return r;
}

/** Group frames by op id (batch-level frames under key "batch"), preserving per-op order. */
export function groupFrames(frames: Frame[]): Record<string, Frame[]> {
  const out: Record<string, Frame[]> = {};
  for (const f of frames) {
    const k = "id" in f ? String(f.id) : "batch";
    (out[k] ??= []).push(f);
  }
  return out;
}
